gitlab/__init__.py

```diff
@@ -28,7 +28,36 @@
     __version__,
 )
 from gitlab.client import Gitlab, GitlabList  # noqa: F401
-from gitlab.const import *  # noqa: F401,F403
+from gitlab.const import (  # noqa: F401
+    DEFAULT_URL,
+    DEVELOPER_ACCESS,
+    GUEST_ACCESS,
+    MAINTAINER_ACCESS,
+    MINIMAL_ACCESS,
+    NO_ACCESS,
+    NOTIFICATION_LEVEL_CUSTOM,
+    NOTIFICATION_LEVEL_DISABLED,
+    NOTIFICATION_LEVEL_GLOBAL,
+    NOTIFICATION_LEVEL_MENTION,
+    NOTIFICATION_LEVEL_PARTICIPATING,
+    NOTIFICATION_LEVEL_WATCH,
+    OWNER_ACCESS,
+    REPORTER_ACCESS,
+    SEARCH_SCOPE_BLOBS,
+    SEARCH_SCOPE_COMMITS,
+    SEARCH_SCOPE_GLOBAL_SNIPPET_TITLES,
+    SEARCH_SCOPE_ISSUES,
+    SEARCH_SCOPE_MERGE_REQUESTS,
+    SEARCH_SCOPE_MILESTONES,
+    SEARCH_SCOPE_PROJECT_NOTES,
+    SEARCH_SCOPE_PROJECTS,
+    SEARCH_SCOPE_USERS,
+    SEARCH_SCOPE_WIKI_BLOBS,
+    USER_AGENT,
+    VISIBILITY_INTERNAL,
+    VISIBILITY_PRIVATE,
+    VISIBILITY_PUBLIC,
+)
```

Comment:
I realize that star imports are awkward, but if this list keeps growing I wonder if they should just be controlled/auto-exported in const's `__all__` kind of how we do in https://github.com/python-gitlab/python-gitlab/blob/2cd15ac44d8a45fa2d0dcab80cc933e3871db7f8/gitlab/v4/objects/__init__.py#L78, but maybe with just specifics exported?

Reply:
So I want to block what gets pulled into the top level `gitlab` namespace. So that is why I am making it explicit.

If we do it in `gitlab.const` then they will still get pulled into the top-level `gitlab` namespace which is what we want to move away from. We want people to use `gitlab.const.FOO`, not `gitlab.FOO`.

The ultimate goal if are able to deprecate is that this will go away at some point.

Does that make sense?

Reply:
To clarify. The reason for this is so that when we add new constants in `gitlab.const` want to make the users access it via `gitlab.const`. So basically this will freeze the variables we bring in from `gitlab.const` into the top-level `gitlab` name-space. Since we would like to move people to access these const variables via `gitlab.const`.

Any new variables/Enums we add will need to be accessed via `gitlab.const`. If we just update the `__all__` value in `gitlab.const` then they will still be able to access them via the top level `gitlab` namespace.

Reply:
> I realize that star imports are awkward, but if this list keeps growing

That is the goal of this. To stop the list from growing. It won't get bigger than this and hopefully if we are able to deprecate it then it will go away in the future. This would be long-term like at least a year.

Reply:
I think I got the idea I was just more wondering about explicitly importing vs explicitly exporting. If deprecating them frpm the main namespace is the goal it would also be easy with `__all__`, e.g. if you try this patch on your branch:

```diff
diff --git a/gitlab/__init__.py b/gitlab/__init__.py
index f1fb7da..e76ed6b 100644
--- a/gitlab/__init__.py
+++ b/gitlab/__init__.py
@@ -19,6 +19,7 @@
 import warnings
 
 import gitlab.config  # noqa: F401
+import gitlab.const
 from gitlab.__version__ import (  # noqa: F401
     __author__,
     __copyright__,
@@ -28,36 +29,18 @@ from gitlab.__version__ import (  # noqa: F401
     __version__,
 )
 from gitlab.client import Gitlab, GitlabList  # noqa: F401
-from gitlab.const import (  # noqa: F401
-    DEFAULT_URL,
-    DEVELOPER_ACCESS,
-    GUEST_ACCESS,
-    MAINTAINER_ACCESS,
-    MINIMAL_ACCESS,
-    NO_ACCESS,
-    NOTIFICATION_LEVEL_CUSTOM,
-    NOTIFICATION_LEVEL_DISABLED,
-    NOTIFICATION_LEVEL_GLOBAL,
-    NOTIFICATION_LEVEL_MENTION,
-    NOTIFICATION_LEVEL_PARTICIPATING,
-    NOTIFICATION_LEVEL_WATCH,
-    OWNER_ACCESS,
-    REPORTER_ACCESS,
-    SEARCH_SCOPE_BLOBS,
-    SEARCH_SCOPE_COMMITS,
-    SEARCH_SCOPE_GLOBAL_SNIPPET_TITLES,
-    SEARCH_SCOPE_ISSUES,
-    SEARCH_SCOPE_MERGE_REQUESTS,
-    SEARCH_SCOPE_MILESTONES,
-    SEARCH_SCOPE_PROJECT_NOTES,
-    SEARCH_SCOPE_PROJECTS,
-    SEARCH_SCOPE_USERS,
-    SEARCH_SCOPE_WIKI_BLOBS,
-    USER_AGENT,
-    VISIBILITY_INTERNAL,
-    VISIBILITY_PRIVATE,
-    VISIBILITY_PUBLIC,
-)
 from gitlab.exceptions import *  # noqa: F401,F403
 
 warnings.filterwarnings("default", category=DeprecationWarning, module="^gitlab")
+
+
+def __getattr__(name):
+    # Deprecate direct access to constants without namespace
+    if name in gitlab.const.__all__:
+        warnings.warn(
+            f"\nDirect access to 'gitlab.{name}' is deprecated and will be"
+            f"\nremoved in a future major python-gitlab release. Please"
+            f"\nuse 'gitlab.const.{name}' instead.",
+            DeprecationWarning)
+        return getattr(gitlab.const, name)
+    raise AttributeError(f"module {__name__} has no attribute {name}")
diff --git a/gitlab/const.py b/gitlab/const.py
index 12faf88..c45494b 100644
--- a/gitlab/const.py
+++ b/gitlab/const.py
@@ -56,3 +56,35 @@ SEARCH_SCOPE_GLOBAL_SNIPPET_TITLES: str = "snippet_titles"
 SEARCH_SCOPE_PROJECT_NOTES: str = "notes"
 
 USER_AGENT: str = f"{__title__}/{__version__}"
+
+
+__all__ = [
+    "DEFAULT_URL",
+    "DEVELOPER_ACCESS",
+    "GUEST_ACCESS",
+    "MAINTAINER_ACCESS",
+    "MINIMAL_ACCESS",
+    "NO_ACCESS",
+    "NOTIFICATION_LEVEL_CUSTOM",
+    "NOTIFICATION_LEVEL_DISABLED",
+    "NOTIFICATION_LEVEL_GLOBAL",
+    "NOTIFICATION_LEVEL_MENTION",
+    "NOTIFICATION_LEVEL_PARTICIPATING",
+    "NOTIFICATION_LEVEL_WATCH",
+    "OWNER_ACCESS",
+    "REPORTER_ACCESS",
+    "SEARCH_SCOPE_BLOBS",
+    "SEARCH_SCOPE_COMMITS",
+    "SEARCH_SCOPE_GLOBAL_SNIPPET_TITLES",
+    "SEARCH_SCOPE_ISSUES",
+    "SEARCH_SCOPE_MERGE_REQUESTS",
+    "SEARCH_SCOPE_MILESTONES",
+    "SEARCH_SCOPE_PROJECT_NOTES",
+    "SEARCH_SCOPE_PROJECTS",
+    "SEARCH_SCOPE_USERS",
+    "SEARCH_SCOPE_WIKI_BLOBS",
+    "USER_AGENT",
+    "VISIBILITY_INTERNAL",
+    "VISIBILITY_PRIVATE",
+    "VISIBILITY_PUBLIC",
+]

```

Reply:
@nejch Thanks. I did (basically) what you suggested. So should be good to go.